File: vidyapath_app/lib/models/event.dart
class EventModel {
  final String id;
  final String name;
  final String? category;
  final String? description;
  final DateTime? eventDate;
  final DateTime? registrationDeadline;
  final Map<String, dynamic>? venue;
  final Map<String, dynamic>? organizer;
  final String? eligibility;
  final String? prizes;
  final double fees;
  final String? status;
  final Map<String, dynamic>? ratings;
  final bool isVerified;
  final DateTime? createdAt;

  EventModel({
    required this.id,
    required this.name,
    this.category,
    this.description,
    this.eventDate,
    this.registrationDeadline,
    this.venue,
    this.organizer,
    this.eligibility,
    this.prizes,
    this.fees = 0,
    this.status,
    this.ratings,
    this.isVerified = false,
    this.createdAt,
  });

  factory EventModel.fromJson(Map<String, dynamic> json) {
    return EventModel(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      category: json['category'],
      description: json['description'],
      eventDate: json['eventDate'] != null ? DateTime.tryParse(json['eventDate']) : null,
      registrationDeadline: json['registrationDeadline'] != null ? DateTime.tryParse(json['registrationDeadline']) : null,
      venue: json['venue'] is Map ? Map<String, dynamic>.from(json['venue']) : null,
      organizer: json['organizer'] is Map ? Map<String, dynamic>.from(json['organizer']) : null,
      eligibility: json['eligibility'],
      prizes: json['prizes'],
      fees: (json['fees'] as num?)?.toDouble() ?? 0,
      status: json['status'],
      ratings: json['ratings'] is Map ? Map<String, dynamic>.from(json['ratings']) : null,
      isVerified: json['isVerified'] ?? false,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
    );
  }

  String get city => venue?['city'] ?? '';
  String get state => venue?['state'] ?? '';
  String get fullAddress => venue?['fullAddress'] ?? '';
  String get organizerName => organizer?['name'] ?? '';
  String get organizerContact => organizer?['contact'] ?? '';
  String get organizerWebsite => organizer?['website'] ?? '';
  double get overallRating => (ratings?['overall'] as num?)?.toDouble() ?? 0.0;
  int get totalReviews => (ratings?['totalReviews'] as num?)?.toInt() ?? 0;
}

File: vidyapath/server/models/Event.js
const mongoose = require('mongoose');

const eventSchema = new mongoose.Schema({
  name: { type: String, required: true, trim: true },
  category: { type: String, enum: ['Sports', 'Cultural', 'Competition', 'Workshop', 'Other'], required: true },
  description: { type: String, default: '' },
  eventDate: Date,
  registrationDeadline: Date,
  venue: {
    city: String, state: String, fullAddress: String,
  },
  location: {
    type: { type: String, enum: ['Point'], default: 'Point' },
    coordinates: { type: [Number], index: '2dsphere' },
  },
  organizer: {
    name: String, contact: String, website: String,
  },
  eligibility: { type: String, default: '' },
  prizes: { type: String, default: '' },
  fees: { type: Number, default: 0 },
  status: { type: String, enum: ['upcoming', 'ongoing', 'completed', 'cancelled'], default: 'upcoming' },
  coverImage: String,
  ratings: {
    overall: { type: Number, default: 0 },
    experience: { type: Number, default: 0 },
    organization: { type: Number, default: 0 },
    value: { type: Number, default: 0 },
    engagement: { type: Number, default: 0 },
    totalReviews: { type: Number, default: 0 },
  },
  customFields: { type: Map, of: mongoose.Schema.Types.Mixed, default: {} },
  isVerified: { type: Boolean, default: false },
  adminUsers: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
  createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  updatedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
}, { timestamps: true });

eventSchema.index({ 'venue.city': 1, 'venue.state': 1 });
eventSchema.index({ category: 1, status: 1 });
eventSchema.index({ eventDate: 1 });

module.exports = mongoose.model('Event', eventSchema);
